src/optimizer/join_elimination.cpp

```diff
@@ -0,0 +1,148 @@
+#include "duckdb/planner/expression/bound_columnref_expression.hpp"
+#include "duckdb/planner/expression/bound_constant_expression.hpp"
+#include "duckdb/planner/expression_iterator.hpp"
+#include "duckdb/planner/operator/logical_aggregate.hpp"
+#include "duckdb/planner/operator/logical_comparison_join.hpp"
+#include "duckdb/planner/operator/logical_distinct.hpp"
+#include "duckdb/planner/operator/logical_filter.hpp"
+#include "duckdb/planner/operator/logical_get.hpp"
+#include "duckdb/planner/operator/logical_order.hpp"
+#include "duckdb/planner/operator/logical_simple.hpp"
+
+namespace duckdb {
+unique_ptr<LogicalOperator> JoinElimination::Optimize(unique_ptr<LogicalOperator> op) {
+	switch (op->type) {
+	case LogicalOperatorType::LOGICAL_ASOF_JOIN:
+	case LogicalOperatorType::LOGICAL_DELIM_JOIN:
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN: {
+		auto &comp_join = op->Cast<LogicalComparisonJoin>();
+		idx_t inner_child_idx;
+		auto join_type = comp_join.join_type;
+		// We only handle the left outer join and right outer join now.
+		if (IsLeftOuterJoin(join_type)) {
+			inner_child_idx = 1;
+		} else if (IsRightOuterJoin(join_type)) {
+			inner_child_idx = 0;
+		} else {
+			break;
+		}
+
+		// Check whether there are used columns for the parent node are came from the inner side.
+		bool all_matched = true;
+		auto &inner_child = comp_join.children[inner_child_idx];
+		const auto all_inner_bindings = inner_child->GetColumnBindings();
+		for (idx_t col_idx = 0; col_idx < all_inner_bindings.size(); col_idx++) {
+			if (column_references.find(all_inner_bindings[col_idx]) != column_references.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+		if (!all_matched) {
+			break;
+		}
+
+		// Check whether all of the unique columns are came from the outer side.
+		auto &outer_child = comp_join.children[1 ^ inner_child_idx];
+		column_binding_set_t all_outer_bindings;
+		for (const auto &binding : outer_child->GetColumnBindings()) {
+			all_outer_bindings.insert(binding);
+		}
+		for (auto &binding : unique_column_references) {
+			if (all_outer_bindings.find(binding) == all_outer_bindings.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+
+		if (all_matched) {
+			op = std::move(outer_child);
+		}
+
+		break;
+	}
+	case LogicalOperatorType::LOGICAL_AGGREGATE_AND_GROUP_BY: {
+		auto &aggr = op->Cast<LogicalAggregate>();
+		for (idx_t col_idx = 0; col_idx < aggr.expressions.size(); col_idx++) {
+			if (aggr.expressions[col_idx]->GetExpressionClass() != ExpressionClass::BOUND_AGGREGATE) {
+				continue;
+			}
+			auto &aggr_expr = aggr.expressions[col_idx]->Cast<BoundAggregateExpression>();
+			// Collect the unique columns.
+			bool has_unique_column = false;
+			if (aggr_expr.IsDistinct()) {
+				has_unique_column = true;
+			} else {
+				auto agg_expr_function_name = aggr_expr.function.name;
+				if (agg_expr_function_name == "max" || agg_expr_function_name == "min") {
+					has_unique_column = true;
```

Comment:
I don't quite follow the logic around `min/max` here, why do `min/max` guarantee uniqueness? e.g.:

```sql
D create table integers(i int, j int);
D insert into integers values (1, 1), (2, 1);
D select max(j) from integers group by i;
┌────────┐
│ max(j) │
│ int32  │
├────────┤
│      1 │
│      1 │
└────────┘
```

Reply:
Yes, `min/max` can not guarantee uniqueness for the specific column. But it can guarantee the uniqueness of one row (It means for each row in the result, the set of all column values for that row is unique). I'll use your case to give some examples:
We have another table called `create table t(a int, b int)`. We have the SQL like `select max(j) from integers left join t on integers.i = t.a  group by i;` If we want to use the `join elimination` rule to get the result like `select max(j) from integers group by i;`. I will simulate the process:

After the left join, we will get the result like this: (`x` means random value, can be `null`)
```
i     j    a    b
1   1    x    x
1   1    x    x   --> same to the previous row
...    -> Maybe more rows have the same values like '1   1    x    x'
2   1    x    x
2   1    x    x  --> same to the previous row
...    -> Maybe more rows have the same values like '2   1    x    x'
```

Because we only need the values for column `i` and column `j`, we can ignore column `a` and column `b`. After that, even though there are many same values rows like '1   1    x    x' and '2   1    x    x'. It may appear a lot of times. But after the `group by and aggregation`, we only need the distinct rows at the end(It means we only need **one row** like`1   1    x    x'  and **one row** like '2   1    x    x'). It means we can the same result at the end, if the join result is like the below:
```
i     j    a    b
1   1    x    x
2   1    x    x
```

So it means for the outer part, if the ancestor has some unique columns which can guarantee the unique rows. The rule will take effect.

It's a little bit hard to understand. So if you have any ideas or thoughts, we can discuss more!

Reply:
There is one thing I forgot to say: if there are a lot of aggregation functions, all of them should be one of the `max/min/distinct`. Otherwise, it can not meet the requirement.

For example, `select max(j) from integers left join t on integers.i = t.a group by i;` can be optimized to `select max(j) from integers group by i;`. But `select max(j), sum(j) from integers left join t on integers.i = t.a group by i;` can not. Because if we have other aggregation functions, the result is different when there are different numbers of the same rows.